lib/AST/RequirementMachine/RewriteSystem.cpp

```diff
@@ -705,10 +718,99 @@ void RewriteSystem::verifyRewriteRules(ValidityPolicy policy) const {
 #undef ASSERT_RULE
 }
 
+std::vector<RequirementError>
+RewriteSystem::getRedundantRequirements() {
+  // The resulting set of redundant requirements.
+  std::vector<RequirementError> redundantRequirements;
+
+  // Map redundant rule IDs to their rewrite path for easy access
+  // in the `isRedundantRule` lambda.
+  llvm::SmallDenseMap<unsigned, RewritePath> redundantRules;
+  for (auto &pair : RedundantRules)
+    redundantRules[pair.first] = pair.second;
+
+  // Collect all rule IDs for each unique requirement ID.
+  llvm::SmallDenseMap<unsigned, llvm::SmallDenseSet<unsigned, 2>>
+      rulesPerRequirement;
+
+  // Collect non-explicit requirements that are not redundant.
+  llvm::SmallDenseSet<unsigned, 2> impliedRequirements;
```

Comment:
Maybe nonExplicitNonRedundantRequirements? 'implied' is kind of vague (I know the comment explains it though)